man/man1/sqlr-query.1
.TH sqlr-query 1 "2002-06-10" "execute a query" SQL\ Relay

.SH NAME
sqlr-query \- execute a query

.SH SYNOPSIS
.B sqlr-query
-host \fIhost\fR -port \fIport\fR -socket \fIsocket\fR -user \fIuser\fR -password \fIpasswd\fR -query \fIquery\fR [-debug]
.br
.B sqlr-query
[-config \fIconfigfile\fR] -id \fIid\fR -query \fIquery\fR [-debug]

.SH DESCRIPTION
The
.B sqlr-query
command executes the query specified in the \fIquery\fR parameter and
returns the result set as a quote/comma/return seperated value list to
standard output.

.SH OPTIONS
.TP
\fB-host\fR \fIhost\fR
hostname of the machine running the listener daemon
.TP
\fB-port\fR \fIport\fR
port number the listener daemon is listening on, for local or remote connections
.TP
\fB-socket\fR \fIsocket\fR
socket the listener daemon is listening on, for local connections
.TP
\fB-user\fR \fIuser\fR
SQL Relay user name (as per config file)
.TP
\fB-password\fR \fIpasswd\fR
SQL Relay password (as per config file)
.TP
\fB-config\fR \fIconfigfile\fR
config file to use (default /etc/sqlrelay/sqlrelay.conf)
.TP
\fB-id\fR \fIid\fR
instance to use (as per config file)
.TP
\fB-query\fR \fIquery\fR
query to be executed
.TP
\fB-debug\fR
produce debugging output

.SH EXAMPLES
.TP
\fBsqlr-query -socket /tmp/examplesocket -user fred -password secret -query "select * from tbl"\fR
Connect to the SQL Relay listener daemon running on the local
machine, with socket \fB/tmp/examplesocket\fR, as user \fBfred\fR,
with password \fBsecret\fR, and run a query.
.TP
\fBsqlr-query -host dbsvr -port 9000 -user fred -password secret -query "select * from tbl" -debug\fR
Connect to the SQL Relay listener daemon running on the remote
machine \fBdbsvr\fR, on port \fB9000\fR, as user \fBfred\fR,
with password \fBsecret\fR, and run a query, producing extra
debugging output.
.TP
\fBsqlr-query -id example -query "select * from tbl"\fR
Connect to the SQL Relay listener daemon for instance \fBexample\fR
(as defined in /etc/sqlrelay/sqlrelay.conf), and run a query.

.SH SEE ALSO
\fBsqlr-fields\fP(1),
\#\fBsqlr-query\fP(1),
\fBsqlrsh\fP(1),
\fBsqlr-cachemanager\fP(8),
\fBsqlr-connection\fP(8),
\fBsqlr-listener\fP(8),
\fBsqlr-scaler\fP(8),
\fBsqlr-start\fP(8),
\fBsqlr-stop\fP(8),
and the documentation in /usr/share/doc/sqlrelay.

.SH AUTHORS
This manual page was written by Andreas Tille <tille@debian.org> and
Florian M. Weps <fweps@pop.agri.ch>, for the Debian GNU/Linux system
(but may be used by others).


NEWS
SQL Relay is a persistent database connection pooling, proxying, throttling, load balancing and query routing/filtering system for Unix and Linux supporting ODBC, Oracle, MySQL, PostgreSQL, Sybase, MS SQL Server, IBM DB2, Firebird, SQLite and MS Access (minimally) with APIs for C, C++, .NET, Perl, Perl-DBI, Python, Python-DB, Zope, PHP, PHP PearDB, Ruby, Ruby-DBI, Java, TCL and Erlang, drop-in replacement libraries for MySQL and PostgreSQL, command line clients and extensive documentation.  The APIs support advanced database operations such as substitution and bind variables, multi-row fetches, client-side result set caching and suspended transactions.  It is ideal for speeding up database-driven web-based applications, accessing databases from unsupported platforms, migrating between databases, distributing access to replicated or clustered databases and throttling database access.


src/cmdline/sqlr-query.cpp
// Copyright (c) 1999-2001  David Muse
// See the file COPYING for more information

#include <config.h>
#include <defaults.h>

#include <sqlrelay/sqlrclient.h>
#include <rudiments/commandline.h>
#include <rudiments/process.h>
#include <sqlrconfigfile.h>

#include <stdio.h>

#ifdef RUDIMENTS_NAMESPACE
using namespace rudiments;
#endif

int main(int argc, const char **argv) {

	#include <version.h>

	sqlrconfigfile	cfgfile;

	commandline	cmdline(argc,argv);

	const char	*config=cmdline.getValue("-config");
	if (!(config && config[0])) {
		config=DEFAULT_CONFIG_FILE;
	}
	const char	*id=cmdline.getValue("-id");
	const char	*host=cmdline.getValue("-host");
	uint16_t	port=charstring::toInteger(
					cmdline.getValue("-port"));
	const char	*socket=cmdline.getValue("-socket");
	const char	*user=cmdline.getValue("-user");
	const char	*password=cmdline.getValue("-password");
	const char	*query=cmdline.getValue("-query");
	uint64_t	rsbs=charstring::toInteger(
				cmdline.getValue("-resultsetbuffersize"));
	if (!rsbs) {
		rsbs=100;
	}
	bool		debug=cmdline.found("-debug");

	if (!(charstring::length(id) ||
		((charstring::length(host) ||
			charstring::length(socket)) &&
				charstring::length(user) &&
				charstring::length(password))) ||
		!(charstring::length(query))) {

		printf("usage: sqlr-query -host host -port port -socket socket -user user -password password -query query [-debug] [-resultsetbuffersize rows]\n"
			"  or   sqlr-query  [-config configfile] -id id -query query [-debug] [-resultsetbuffersize rows]\n");
		process::exit(1);
	}

	if (charstring::length(id) && cfgfile.parse(config,id)) {

		// get the host/port/socket/username/password
		host="localhost";
		port=cfgfile.getPort();
		socket=cfgfile.getUnixPort();
		// FIXME: this can return 0
		usercontainer	*currentnode=NULL;
		cfgfile.getUserList()->getDataByIndex(0,&currentnode);
		user=currentnode->getUser();
		password=currentnode->getPassword();
	}

	sqlrconnection	sqlrcon(host,port,socket,user,password,0,1);
	sqlrcursor	sqlrcur(&sqlrcon);

	if (debug) {
		sqlrcon.debugOn();
	}

	int32_t	exitval=0;
	sqlrcur.dontGetColumnInfo();
	sqlrcur.setResultSetBufferSize(rsbs);
	if (sqlrcur.sendQuery(query)) {
		uint64_t	i=0;
		uint32_t	cols=sqlrcur.colCount();
		const char	*field="";
		while (cols && field) {
			for (uint32_t j=0; j<cols; j++) {
				if ((field=sqlrcur.getField(i,j))) {
					printf("\"%s\"",field);
					if (j<cols-1) {
						printf(",");
					} else {
						printf("\n");
					}
				}
			}
			i++;
		}
	} else {
		printf("%s\n",sqlrcur.errorMessage());
		exitval=1;
	}
	sqlrcon.endSession();

	process::exit(exitval);
}
